wdl-lsp/src/lib.rs

```diff
@@ -0,0 +1,485 @@
+            let analyzer_client = client.clone();
+            let analyzer_workspace = workspace.clone();
+
+            Self {
+                client,
+                options,
+                analyzer: Analyzer::new_with_changes(
+                    move |kind, current, total, context| {
+                        Self::report_analysis_progress(
+                            analyzer_client.clone(),
+                            context.clone(),
+                            kind,
+                            current,
+                            total,
+                        )
+                    },
+                    move || validator(lint),
+                    move |uri| changes(&analyzer_workspace, uri),
+                ),
+
+                workspace,
+            }
+        });
+
+        let stdin = tokio::io::stdin();
+        let stdout = tokio::io::stdout();
+        tower_lsp::Server::new(stdin, stdout, socket)
+            .serve(service)
+            .await;
+
+        Ok(())
+    }
+
+    /// Registers a generic watcher for all files/directories in the workspace.
+    async fn register_watcher(&self) {
+        self.client
+            .register_capability(vec![Registration {
+                id: Uuid::new_v4().to_string(),
+                method: "workspace/didChangeWatchedFiles".into(),
+                register_options: Some(
+                    to_value(DidChangeWatchedFilesRegistrationOptions {
+                        watchers: vec![FileSystemWatcher {
+                            // We use a generic glob so we can be notified for when directories,
+                            // which might contain WDL documents, are deleted
+                            glob_pattern: GlobPattern::String("**/*".to_string()),
+                            kind: None,
+                        }],
+                    })
+                    .expect("should convert to value"),
+                ),
+            }])
+            .await
+            .expect("failed to register capabilities with client");
+    }
+
+    /// Starts an analysis task by creating a work done token.
+    ///
+    /// If the client returned an error during token creation, `None` is
+    /// returned.
+    async fn start_analysis_task(&self) -> Option<String> {
+        let token = Uuid::new_v4().to_string();
+        if self
+            .client
+            .send_request::<WorkDoneProgressCreate>(WorkDoneProgressCreateParams {
+                token: NumberOrString::String(token.clone()),
+            })
+            .await
+            .is_err()
+        {
+            return None;
+        }
+
+        self.client
+            .send_notification::<Progress>(ProgressParams {
+                token: NumberOrString::String(token.clone()),
+                value: ProgressParamsValue::WorkDone(WorkDoneProgress::Begin(
+                    WorkDoneProgressBegin {
+                        title: "Sprocket".to_string(),
```

Comment:
Since this technically _isn't_ sprocket yet, how do we want to handle this?

Reply:
I think `wdl-lsp` is more appropriate

Reply:
Or the more flexible `env!("CARGO_CRATE_NAME")`? (copied from a bit further down)

Reply:
Agreed, this should be coming from the sever options as a "display name" for the server.

We'll default to the crate name if not specified.